### PR TITLE
add support for --review-pr

### PR DESCRIPTION
**Pull request by boegel**
wrapper PR for @nudded's #1005

It contains a bunch of cleanups on top of @nudded's work, together with code style cleanup.

@stdweird, @wpoely86: please review/test

### CODE REVIEW

**Comment by boegel** [comment]
No additional unit tests yet, but I'll look into that later.

File: easybuild/tools/github.py [hunk 1]
```diff
@@ -190,33 +190,107 @@ class GithubError(Exception):
...
+    Fetch latest SHA1 for a specified repository and branch.
+    @param repo: GitHub repository
+    @param account: GitHub account
+    @param branch: branch to fetch latest SHA1 for
+    @return: latest SHA1
+    """
+    status, data = github_api_get_request(lambda x: x.repos[account][repo].branches)
+    if not status == HTTP_STATUS_OK:
+        tup = (branch, account, repo, status, data)
+        _log.error("Failed to get latest commit sha for branch %s from %s/%s (status: %d %s)" % tup)
```
Comment (stdweird) [inline easybuild/tools/github.py]: error message doesn refelct the code. there is nothing `sha` releated here

File: easybuild/tools/github.py [hunk 2]
```diff
@@ -190,33 +190,107 @@ class GithubError(Exception):
...
+
+    res = None
+    for entry in data:
+        if entry[u'name'] == branch:
+            res = entry['commit']['sha']
```
Comment (stdweird) [inline easybuild/tools/github.py]: is it intentionally redefining `res`? (this is latest first, and `data` is sorted how?)

File: easybuild/framework/easyconfig/tools.py [hunk 3]
```diff
@@ -248,13 +253,23 @@ def det_easyconfig_paths(orig_paths, from_pr=None, easyconfigs_pkg_paths=None):
...
         pr_files = fetch_easyconfigs_from_pr(from_pr)
-        ec_files = [path for path in pr_files if path.endswith('.eb')]
 
-    elif ec_files and easyconfigs_pkg_paths:
+        if ec_files:
+            # replace paths for specified easyconfigs that are touched in PR
+            for i, ec_file in enumerate(ec_files):
+                for pr_file in pr_files:
+                    if ec_file == os.path.basename(pr_file):
+                        ec_files[i] = pr_file
```
Comment (stdweird) [inline easybuild/framework/easyconfig/tools.py]: add a debug log here, this is a very tricky loop+if block

File: easybuild/framework/easyconfig/tools.py [hunk 4]
```diff
@@ -292,7 +307,7 @@ def det_easyconfig_paths(orig_paths, from_pr=None, easyconfigs_pkg_paths=None):
     return [(ec_file, False) for ec_file in ec_files]
 
 
-def parse_easyconfigs(paths):
+def parse_easyconfigs(paths, validate=True):
```
Comment (stdweird) [inline easybuild/framework/easyconfig/tools.py]: use `**kwargs`, if you are going to pass it as-is to `process_easyconfig`  (and add it to it the docstring)

File: easybuild/framework/easyconfig/tools.py [hunk 5]
```diff
@@ -333,3 +348,100 @@ def stats_to_str(stats):
...
+
+    Matching versionsuffix is considered prior to any versionsuffix.
+
+    Exact software version is considered prior to matching major/minor version numbers,
+    and only matching major version number. Any software version is considered last.
+
+    The following criteria are considered, in order (with 'version criterion' being either an
+    exact version match, a major/minor version match, a major version match, or no version match).
+
+    (i)   software version criterion, matching versionsuffix and toolchain name/version
```
Comment (stdweird) [inline easybuild/framework/easyconfig/tools.py]: so they all have `software version criterion`. don't copy paste, refactor.

File: easybuild/framework/easyconfig/tools.py [hunk 6]
```diff
@@ -333,3 +348,100 @@ def stats_to_str(stats):
...
+    Software version is considered more important than toolchain.
+
+    Toolchain is considered with exact same version prior to without version (only name).
+
+    Matching versionsuffix is considered prior to any versionsuffix.
+
+    Exact software version is considered prior to matching major/minor version numbers,
+    and only matching major version number. Any software version is considered last.
+
+    The following criteria are considered, in order (with 'version criterion' being either an
```
Comment (stdweird) [inline easybuild/framework/easyconfig/tools.py]: so which is it: the previous block of text or the list below?

File: easybuild/framework/easyconfig/tools.py [hunk 7]
```diff
@@ -333,3 +348,100 @@ def stats_to_str(stats):
...
+    if len(parsed_version) >= 2:
+        version_patterns.append(r'%s\.%s\.[0-9_A-Za-z]+' % tuple(parsed_version[:2]))  # major/minor version match
+    if parsed_version != parsed_version[0]:
+        version_patterns.append(r'%s\.[0-9-]+\.[0-9_A-Za-z]+' % parsed_version[0])  # major version match
+    version_patterns.append(r'[0-9._A-Za-z-]+')  # any version
+
+    regexes = []
+    for version_pattern in version_patterns:
+        regexes.extend([
+            re.compile((r"^\S+/%s-%s%s%s.eb$" % (name, version_pattern, toolchain_pattern, versionsuffix))),
```
Comment (stdweird) [inline easybuild/framework/easyconfig/tools.py]: do the `re.compile` in a separate step on the whole list of patterns (use `map`).
make a pattern that holds all the common pattern `r"^\S+/%s-%s%%s.eb$" %(name, version_pattern)` is a good start

File: easybuild/framework/easyconfig/tools.py [hunk 8]
```diff
@@ -333,3 +348,100 @@ def stats_to_str(stats):
...
+        toolchain_name_pattern = ''
+        toolchain_pattern = ''
+
+    potential_paths = [glob.glob(ec_path) for ec_path in create_paths(path, name, '*')]
+    potential_paths = sum(potential_paths, [])  # flatten
+
+    parsed_version = LooseVersion(version).version
+    version_patterns = [version]  # exact version match
+    if len(parsed_version) >= 2:
+        version_patterns.append(r'%s\.%s\.[0-9_A-Za-z]+' % tuple(parsed_version[:2]))  # major/minor version match
```
Comment (stdweird) [inline easybuild/framework/easyconfig/tools.py]: `[0-9_A-Za-z]` is `\w`, please use that (everywhere)

File: easybuild/framework/easyconfig/tools.py [hunk 9]
```diff
@@ -333,3 +348,100 @@ def stats_to_str(stats):
...
+    if parsed_version != parsed_version[0]:
+        version_patterns.append(r'%s\.[0-9-]+\.[0-9_A-Za-z]+' % parsed_version[0])  # major version match
```
Comment (stdweird) [inline easybuild/framework/easyconfig/tools.py]: `0-9` is `\d`

File: easybuild/framework/easyconfig/tools.py [hunk 10]
```diff
@@ -333,3 +348,100 @@ def stats_to_str(stats):
...
+    ecs, _ = parse_easyconfigs([(fp, False) for fp in pr_files], validate=False)
+    for ec in ecs:
+        files = find_related_easyconfigs(repo_path, ec['ec'])
+        _log.debug("File in PR#%s %s has these related easyconfigs: %s" % (pr, ec['spec'], files))
+        if files:
+            diff = multidiff(ec['spec'], files, colored=colored)
+            msg = diff
+        else:
+            msg = "\n(no related easyconfigs found for %s)\n" % os.path.basename(ec['spec'])
+        print(msg)
```
Comment (stdweird) [inline easybuild/framework/easyconfig/tools.py]: generate all the text, print it all once

Comment (stdweird) [inline easybuild/framework/easyconfig/tools.py]: and print it all at once in main

File: easybuild/tools/github.py [hunk 11]
```diff
@@ -175,8 +175,8 @@ def read(self, path, api=True):
         # https://raw.github.com/hpcugent/easybuild/master/README.rst
         if not api:
             outfile = tempfile.mkstemp()[1]
-            url = ("http://raw.github.com/%s/%s/%s/%s" % (self.githubuser, self.reponame, self.branchname, path))
-            urllib.urlretrieve(url, outfile)
+            url = ("%s/%s/%s/%s/%s" % (GITHUB_RAW, self.githubuser, self.reponame, self.branchname, path))
```
Comment (stdweird) [inline easybuild/tools/github.py]: `'/'.join` is broken?

File: easybuild/tools/options.py [hunk 12]
```diff
@@ -175,6 +175,7 @@ def override_options(self):
             'allow-modules-tool-mismatch': ("Allow mismatch of modules tool and definition of 'module' function",
                                             None, 'store_true', False),
             'cleanup-builddir': ("Cleanup build dir after successful installation.", None, 'store_true', True),
+            'color': ("Allow color output", None, 'store_true', True),
```
Comment (stdweird) [inline easybuild/tools/options.py]: color output for what?

File: easybuild/tools/utilities.py [hunk 13]
```diff
@@ -118,3 +121,12 @@ def import_available_modules(namespace):
...
                 modules.append(__import__(modpath, globals(), locals(), ['']))
     return modules
+
+
+def det_terminal_size():
+    """
+    Determine the current size of the terminal window.
+    @return: tuple with terminal width and height
+    """
+    height, width, _, _ = struct.unpack('HHHH', fcntl.ioctl(0, termios.TIOCGWINSZ, struct.pack('HHHH', 0, 0, 0, 0)))
```
Comment (stdweird) [inline easybuild/tools/utilities.py]: where did you get this?

**Comment by stdweird** [comment]
@boegel didn't check any of the tyests, have to rereview the multidiff (makes my head hurt :smile: )

**Comment by boegel** [comment]
I'll need to reopen this to target `develop`, crap...

**Comment by boegel** [comment]
PR targeting `develop` at #1383

**Pull request closed without merging**